crates/red_knot_python_semantic/src/semantic_index/builder.rs

```diff
@@ -869,7 +827,16 @@ where
             range: _,
         }) = pattern
         {
-            self.add_symbol_and_definition(name);
+            let symbol = self.add_or_update_symbol(name.id().clone(), SymbolFlags::IS_DEFINED);
+            let state = self.pattern_state.as_ref().unwrap();
```

Comment:
This includes a bunch of `unwrap`s similar to this. Let me think about if there's a way to avoid them.

Reply:
> I guess this doesn't work because the borrow checker will get all mad at you because we call other methods on `self`.
> 
> My normal solution would be to just `take` the value when entering the method and restore it when exiting, but I think that doesn't work as well because the value must be set for `walk_pattern` in case there are nested patterns.

Correct on both points.

Reply:
Another solution would be to override the `visit_pattern` to accept the state as a parameter but that means that the `walk_pattern` logic needs to be inlined as well to pass in the parameter.